dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h

```diff
@@ -0,0 +1,359 @@
+        try {
+          const auto image = ImageFactory::CreateImage(
+                              static_cast<const uint8 *>(input_data), in_size);
+          const auto dims = image->GetImageDims();
+          info.heights[0] = std::get<0>(dims);
+          info.widths[0] = std::get<1>(dims);
+          info.nvjpeg_support = false;
+          output_info_[i] = info;
+        } catch (const std::runtime_error &e) {
+          DALI_FAIL(e.what() + "File: " + file_name);
+        }
+      } else {
+        if (ShouldBeHybrid(info)) {
+          image_decoders_[i] = decoder_huff_hybrid_;
+          image_states_[i] = decoder_huff_hybrid_state_[i];
+        } else {
+          image_decoders_[i] = decoder_huff_host_;
+          image_states_[i] = decoder_host_state_[i];
+        }
+
+        /*
+        // TODO(spanev): add this function when integrating Crop
+        nvjpegnvjpegDecodeParamsSetROI(decode_params_[pos], offset_x, offset_y, roi_w, roi_h);
+        */
+      }
+      const auto c = static_cast<Index>(NumberOfChannels(output_image_type_));
+      output_shape_[i] = Dims({info.heights[0], info.widths[0], c});
+      output_info_[i] = info;
+    }
+  }
+
+  void ProcessImages(MixedWorkspace* ws) {
+    // Creating output shape and setting the order of images so the largest are processed first
+    // (for load balancing)
+    std::vector<std::pair<size_t, size_t>> image_order(batch_size_);
+    for (int i = 0; i < batch_size_; i++) {
+      image_order[i] = std::make_pair(volume(output_shape_[i]), i);
+    }
+    std::sort(image_order.begin(), image_order.end(),
+              std::greater<std::pair<size_t, size_t>>());
+
+    auto& output = ws->Output<GPUBackend>(0);
+    TypeInfo type = TypeInfo::Create<uint8_t>();
+    output.set_type(type);
+    output.Resize(output_shape_);
+    output.SetLayout(DALI_NHWC);
+
+    CUDA_CALL(cudaEventRecord(master_event_, ws->stream()));
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaStreamWaitEvent(streams_[i], master_event_, 0));
+    }
+
+    for (int idx = 0; idx < batch_size_; ++idx) {
+      const int i = image_order[idx].second;
+
+      const auto &in = ws->Input<CPUBackend>(0, i);
+      const auto file_name = in.GetSourceInfo();
+      cudaStream_t stream = ws->stream();
+      thread_pool_.DoWorkWithID(std::bind(
+            [this, i, file_name, stream, &in, &output](int idx, int tid) {
+              SampleWorker(i, file_name, in.size(), tid,
+                              in.data<uint8_t>(), output.mutable_tensor<uint8_t>(i));
+            }, i, std::placeholders::_1));
+    }
+
+    thread_pool_.WaitForWork();
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaEventRecord(events_[i], streams_[i]));
+      CUDA_CALL(cudaStreamWaitEvent(ws->stream(), events_[i], 0));
+    }
+  }
+
+  void SampleWorker(int i, string file_name, int in_size, int tid,
+                    const uint8_t* input_data, uint8_t* output_data) {
```

Comment:
Please add documentation, about what this function should do.
Also it would be nice to use more meaningful names for variables, than `i` and `tid` ;)

Reply:
Done